Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/DefaultPrimaryPointerSelector.cs

```diff
@@ -0,0 +1,156 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System.Collections.Generic;
+
+namespace Microsoft.MixedReality.Toolkit.Input
+{
+    /// <summary>
+    /// Default primary pointer selector. The primary pointer is chosen among all interaction enabled ones using the following rules in order:
+    ///   1. Currently pressed pointer that has been pressed for the longest
+    ///   2. Pointer that was released most recently
+    ///   3. Pointer that became interaction enabled most recently
+    /// </summary>
+    public class DefaultPrimaryPointerSelector : IMixedRealityPrimaryPointerSelector, IMixedRealityPointerHandler
+    {
+        private readonly Dictionary<IMixedRealityPointer, PointerInfo> pointerInfos = new Dictionary<IMixedRealityPointer, PointerInfo>();
+
+        public void Initialize()
+        {
+            IMixedRealityInputSystem inputSystem;
+            if (MixedRealityServiceRegistry.TryGetService<IMixedRealityInputSystem>(out inputSystem))
+            {
+                inputSystem.RegisterHandler<IMixedRealityPointerHandler>(this);
+            }
+        }
+
+        public void Destroy()
+        {
+            IMixedRealityInputSystem inputSystem;
+            if (MixedRealityServiceRegistry.TryGetService<IMixedRealityInputSystem>(out inputSystem))
+            {
+                inputSystem.UnregisterHandler<IMixedRealityPointerHandler>(this);
+            }
+        }
+
+        #region IMixedRealityPrimaryPointerSelector
+
+        public void RegisterPointer(IMixedRealityPointer pointer)
+        {
+            var pointerInfo = new PointerInfo(pointer);
+            pointerInfos.Add(pointer, pointerInfo);
+        }
+
+        public void UnregisterPointer(IMixedRealityPointer pointer)
+        {
+            pointerInfos.Remove(pointer);
+        }
+
+        virtual public IMixedRealityPointer Update()
+        {
+            IMixedRealityPointer primaryPointer = null;
+            PointerInfo primaryInfo = null;
+
+            foreach (var keyValue in pointerInfos)
+            {
+                var pointer = keyValue.Key;
+                var info = keyValue.Value;
+                info.Update(pointer);
+
+                if (info.IsInteractionEnabled && 
+                    (primaryInfo == null || 
+                    (info.IsPressed && (!primaryInfo.IsPressed || info.PressedTimestamp < primaryInfo.PressedTimestamp)) ||
+                    (!primaryInfo.IsPressed && info.ReleasedTimestamp > primaryInfo.ReleasedTimestamp)))
+                {
```

Comment:
It might be interesting to add tests for this - these things are somewhat hard to validate manually, and changing this in the future might be scary because validating manually is so hard.

However, manually newing this up in a test and then driving this with calls to OnPointerDown/OnPointerUp (or other mock pointer objects) should be enough to have some basic level of testing for this.

Reply:
Added!

Reply:
This is great, thanks!